Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -49,23 +49,52 @@ extension HTTPClient {
             @inlinable
             func writeChunks<Bytes: Collection>(of bytes: Bytes, maxChunkSize: Int) -> EventLoopFuture<Void>
             where Bytes.Element == UInt8 {
-                let iterator = UnsafeMutableTransferBox(bytes.chunks(ofCount: maxChunkSize).makeIterator())
-                guard let chunk = iterator.wrappedValue.next() else {
+                // `StreamWriter` is has design issues, for example
+                // - https://github.com/swift-server/async-http-client/issues/194
+                // - https://github.com/swift-server/async-http-client/issues/264
+                // - We're not told the EventLoop the task runs on and the user is free to return whatever EL they
+                //   want.
+                // One important consideration then is that we must lock around the iterator because we could be hopping
+                // between threads.
+                let iterator = NIOLockedValueBox(
+                    bytes.chunks(ofCount: maxChunkSize).enumerated().makeIterator()
+                )
+                guard let chunk = iterator.withLockedValue({ $0.next() }) else {
                     return self.write(IOData.byteBuffer(.init()))
                 }
 
                 @Sendable  // can't use closure here as we recursively call ourselves which closures can't do
-                func writeNextChunk(_ chunk: Bytes.SubSequence) -> EventLoopFuture<Void> {
-                    if let nextChunk = iterator.wrappedValue.next() {
-                        return self.write(.byteBuffer(ByteBuffer(bytes: chunk))).flatMap {
-                            writeNextChunk(nextChunk)
-                        }
+                func writeNextChunk(
+                    _ chunk: (offset: Int, element: ChunksOfCountCollection<Bytes>.Element),
+                    allDone: EventLoopPromise<Void>
+                ) {
+                    if let nextElement = iterator.withLockedValue({ $0.next() }) {
+                        let (index, chunk) = nextElement
+                        self.write(.byteBuffer(ByteBuffer(bytes: chunk))).map {
+                            if (index + 1) % 4 == 0 {
+                                // Let's not stack-overflow if the futures insta-complete which they at least in HTTP/2
+                                // mode.
+                                // Also, we must frequently return to the EventLoop because we may get the pause signal
+                                // from another thread. If we fail to do that promptly, we may balloon our body chunks
+                                // into memory.
+                                allDone.futureResult.eventLoop.execute {
```

Comment:
This is the crucial line (apart from the thread unsafety from `UnsafeMutableTransferBox`)